rfd/0143-external-k8s-joining.md

```diff
@@ -0,0 +1,222 @@
+
+It's important to note that this join method will be less secure than the 
+`kubernetes` join method. This should be explicitly mentioned in documentation
+for this join method, and encourage the use of `kubernetes` where possible.
+
+This is because the `kubernetes` join method makes use of the TokenReview 
+endpoint which performs additional checks to the validity of the token. This 
+includes checking that the pod and service account listed within the JWT claims 
+exist and are running. This increases the complexity of falsifying a 
+token and also ensures that a token cannot be used beyond the lifetime of the 
+pod it is bound to.
+
+However, this join method is still more secure than the long-lived secret
+based `token` join method that operators are forced to use due to this
+external Kubernetes join method not existing.
+
+The risk of token reuse will be mitigated by the inclusion of a nonce within
+the audience of the JWT, as described under details.
+
+## Alternatives
+
+### Extending the existing `kubernetes` join method
+
+Extending the existing `kubernetes` join method to support external joining
+has two key issues.
+
+The first issue is connectivity. The `kubernetes` join method relies on the Auth
+Service's ability to call the TokenReview endpoint exposed by the Kubernetes API
+server. This means there must be network connectivity and the appropriate 
+firewall rules to allow the request to pass. Many users operate clusters that
+are not exposed to the internet, and this means Teleport Cloud would not work
+without some additional way to pass traffic to the external Kubernetes cluster
+API server.
+
+The second issue is authentication. Calling the TokenReview RPC requires some 
+form of authentication. At the moment, the Auth Service relies on a Kubernetes
+service account available to it by virtue of being a workload within the
+cluster.
+
+The issue with authentication could be solved by allowing an operator to
+configure credentials for the Kubernetes cluster as part of the join token
+specification. This, however, does not solve the issue of connectivity and
+introduces further complexities as to how we would safeguard customer
+credentials in Teleport Cloud.
+
+One solution that would solve the connectivity and authentication problems
+would be to use an existing Teleport Kubernetes Agent deployed into a
+Kubernetes Cluster as a "stepping stone". Existing methods for executing a
+Kubernetes request could be leveraged by the Auth Server and identity
+impersonation used to ensure the request is completed with a group that has
+access to the TokenReview RPC.
+
+This does raise a few concerns:
+
+- This method would not be suitable for deploying the first Kubernetes Agent
+  into a cluster.
+- This raises the potential fallout from a compromised Kubernetes Agent. A
+  compromised agent could be used to impersonate any service account.
+- It is relatively unusual in the Teleport codebase for the Auth Server to
```

Comment:
There are no other use cases where auth dials to proxy to dial to any other agent.

One possible solution for this would be the joining method to terminate in the Teleport proxy, proxy calls the agent through the reverse tunnel, the agent checks the validity of the token, and finally proxy sends to the Auth server the payload and the proof of validation.

This would require that the joining service knows the name of the Kubernetes cluster they are running in and that name matches with a cluster in Teleport. 

There are also cases where the agent doesn't live inside the cluster - Kubernetes auto-discovery - and for those cases, it's still possible to use that but the setup of permissions to do the token review adds an extra layer of complexity because it involves changing the current required permissions and EKS forces you to change the cluster role you created for each cluster - there is no helm chart upgrade for that :cry:

Reply:
081a4f650d